src/modvault/modmodel.py

```diff
@@ -0,0 +1,186 @@
+        item_index = self._itemlist.index(item)
+        index = self.index(item_index, 0)
+        self.dataChanged.emit(index, index)
+
+
+class ModSortModel(QSortFilterProxyModel):
+    class SortType(Enum):
+        ALPHABETICAL = 0
+        DATE = 1
+        RATING = 2
+        DOWNLOADS = 3
+
+    def __init__(self, model):
+        QSortFilterProxyModel.__init__(self)
+        self._sort_type = self.SortType.RATING
+        self.setSourceModel(model)
+        self.sort(0)
+
+    def lessThan(self, leftIndex, rightIndex):
+        left = self.sourceModel().data(leftIndex, Qt.DisplayRole).mod
+        right = self.sourceModel().data(rightIndex, Qt.DisplayRole).mod
+
+        comp_list = [self._lt_type, self._lt_fallback]
+
+        for lt in comp_list:
+            if lt(left, right):
+                return True
+            elif lt(right, left):
+                return False
+        return False
+
+    def _lt_type(self, left, right):
+        stype = self._sort_type
+        stypes = self.SortType
+
+        if stype == stypes.ALPHABETICAL:
+            if left.name.lower() == right.name.lower():
+                return left.version > right.version
+            return left.name.lower() < right.name.lower()
+        elif stype == stypes.DATE:
+            if left.date == right.date:
+                return left.name.lower() < right.name.lower()
+            return left.date > right.date
+        elif stype == stypes.RATING:
+            if left.likes == right.likes:
+                return left.date > right.date
+            return left.likes > right.likes
+        elif stype == stypes.DOWNLOADS:
+            if left.downloads == right.downloads:
+                return left.date > right.date
+            return left.downloads > right.downloads
+
+    def _lt_fallback(self, left, right):
+        return left.uid < right.uid
+
+    @property
+    def sort_type(self):
+        return self._sort_type
+
+    @sort_type.setter
+    def sort_type(self, stype):
+        self._sort_type = stype
+        self.invalidate()
+
+    def filterAcceptsRow(self, row, parent):
+        index = self.sourceModel().index(row, 0, parent)
+        if not index.isValid():
+            return False
+        mod = index.data().mod
+
+        return self.filter_accepts_mod(mod)
+
+    def filter_accepts_mod(self, mod):
+        return True
+
+
+class ModFilterModel(ModSortModel):
+    class FilterType(Enum):
+        ALL = 0
+        UI = 1
+        SIM = 2
+        YOURS = 3
+        INSTALLED = 4
+
+    def __init__(self, model):
+        self._filter_type = self.FilterType.ALL
+        self._search_str = ""
+        ModSortModel.__init__(self, model)
+
+    def filter_accepts_mod(self, mod):
+        search_str = self._search_str
+        if search_str != "":
+            if not (mod.author.lower().find(search_str) != -1 or mod.name.lower().find(search_str) != -1 or
+                    mod.description.lower().find(" " + search_str + " ") != -1):
```

Comment:
This will fail to find a word if it's at the beginning or end of the description, or if there's punctuation around it. Using the 're' module and the word boundary character '\b' should be robust enough.

Reply:
I didn't touch or look in to that. I'll have a look.

Reply:
We could add more to include words at begining and end, or make it easy and remove both " ".

Reply:
It might actually be better, since people might want to search by parts of words.

Reply:
I guess the restriction to a word for search in description was, that a search in lengthy description would lead to too many unwanted search results. But let's keep it simple .. or maybe have a checkbox 'include description'.

Reply:
(you're aware that you answered my 'X or Y ?' with kinda 'yes' ;-)